ledger_device_sdk/src/io_new.rs

```diff
@@ -0,0 +1,650 @@
+        DecodedEventType::Apdu {
+            header,
+            offset,
+            length,
+        } => {
+            comm.pending_apdu = true;
+            comm.pending_header = header;
+            comm.pending_offset = offset;
+            comm.pending_length = length;
+            return true;
+        }
+        _ => {}
+    }
+    false
+}
+
+fn fetch_apdu_header_impl<const N: usize>() -> Option<ApduHeader> {
+    let comm = unsafe { get_comm::<N>() };
+    if comm.pending_apdu {
+        Some(comm.pending_header)
+    } else {
+        None
+    }
+}
+
+fn reply_status_impl<const N: usize>(reply: Reply) {
+    let comm = unsafe { get_comm::<N>() };
+    if comm.pending_apdu {
+        comm.pending_apdu = false;
+    }
+    let _ = comm.begin_tx().send(reply);
+}
+
+// BOLOS APDU Handling
+fn handle_bolos_apdu<const N: usize>(comm: &mut Comm<N>, ins: u8) {
+    match ins {
+        // Get Information INS: retrieve App name and version
+        0x01 => {
+            let mut tx = comm.begin_tx();
+            let _ = tx.append(&[0x01]);
+            const MAX_TAG_LENGTH: u8 = 32; // maximum length for the buffer containing app name/version.
+            let mut tag_buf = [0u8; MAX_TAG_LENGTH as usize];
+
+            // ---- App name ----
+
+            let name_len = unsafe {
+                os_registry_get_current_app_tag(
+                    BOLOS_TAG_APPNAME,
+                    tag_buf.as_mut_ptr(),
+                    MAX_TAG_LENGTH as u32,
+                )
+            };
+
+            if name_len > MAX_TAG_LENGTH.into() {
+                let _ = tx.send(StatusWords::Panic); // this should never happen
+                return;
+            }
+
+            let _ = tx.append(&[name_len as u8]);
+            let _ = tx.append(&tag_buf[..name_len as usize]);
+
+            // ---- App version ----
+
+            let ver_len = unsafe {
+                os_registry_get_current_app_tag(
+                    BOLOS_TAG_APPVERSION,
+                    tag_buf.as_mut_ptr(),
+                    MAX_TAG_LENGTH as u32,
+                )
+            };
+
+            if ver_len > MAX_TAG_LENGTH.into() {
+                let _ = tx.send(StatusWords::Panic); // this should never happen
+                return;
+            }
+
+            let _ = tx.append(&[ver_len as u8]);
+            let _ = tx.append(&tag_buf[..ver_len as usize]);
+
+            // ---- Flags ----
+            let flags_byte = unsafe { os_flags() } as u8;
+            // flags length (always 1 currently) then flags byte
+            let _ = tx.append(&[1]);
+            let _ = tx.append(&[flags_byte]);
+            let _ = tx.send(StatusWords::Ok);
+        }
+        // Quit Application INS
+        0xa7 => {
+            let _ = comm.begin_tx().send(StatusWords::Ok);
+            crate::exit_app(0);
+        }
+        // Unknown INS within BOLOS namespace
+        _ => {
+            let _ = comm.begin_tx().send(StatusWords::BadIns);
+        }
+    }
+}
```

Comment:
A more general comment: it could be relvant to split the module into submodule to ease readability and maintenance. The new io module is far better in terms of design and architecture but it is still a big file of several hundred of lines of code. WDYT ?